### DISCUSSION TITLE
WebOne on ARMv6

### DISCUSSION THREAD

**Discussion by way5** [atauenis/webone#75] [General]
Hi everyone,

just wanted to share my deepest frustration with MS's dotnet. :-) 
The problem is that WebOne can NOT be compiled for ARMv6 cpus (ex. RPI Zero) because of known incompatibility (https://github.com/actions/runner/issues/688), that I've just came along looking for a solution for Seg Faults while trying to run dotnet or WebOne on raspberry.

I hope this info might be useful for someone in order to save some time having similar problem.

Thank you microsoft...

**Comment by way5** [comment]
@atauenis, I also want to suggest to add a few directives into webone.service in order to prevent "Start request repeated too quickly." if an error emerged during the service startup. 

```
Restart=on-failure
...
StartLimitBurst=2
StartLimitInterval=30
```

that may be changed as you think it more suitable.

**Reply by atauenis** [reply]
I think that if WebOne can't start at all, any time and attempt limits won't help. The error will be same, `Start request repeated too quickly` + `Failed with result 'start-limit-hit'`.

There are few known cases where there may be failures on WebOne start:
1. Problems with ELF file (architecture mismatch like this).
2. Problems with .NET.
3. Problems with loading `/etc/webone.conf`. In this case systemd status log will contain error message.
4. Problems with opening log file for write. Also systemd status log will have the error message. I've tested this when worked on implementing the systemd support.
5. Problems on opening TCP port for listen. Then `/var/log/webone.log` will contain the error message. And even in this case default systemd timeouts correctly stopping the service.

**Comment by atauenis** [comment]
Seems that some kind of ARMv6 support were added in .NET 7 & 8.
https://github.com/dotnet/runtime/issues/7764#issuecomment-1043468998
https://github.com/dotnet/runtime/blob/v8.0.0/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L2039
https://stackoverflow.com/questions/74561337/can-i-publish-targeting-linux-armv6-runtime-in-net-core-7
And now there's chances to build a working version of WebOne for Raspberry Pi Zero/1. At this moment I haven't tried yet.

**Reply by atauenis** [reply]
Yes! It's possible to build a working binaries for ARMv6-based Raspberry Pi. Used script and Docker image [linked above](https://stackoverflow.com/questions/74561337/can-i-publish-targeting-linux-armv6-runtime-in-net-core-7), and successfully built working version for ARMv6. The only thing that is impossible - build deb/rpm packages. The utilities I've using to do this at this moment supports .NET 6.0 maximum, and seems to be abandoned by developer. But, may be, sometimes they will be updated. I've waited for .NET 6 support for approximately a year too, and it became.

**[WebOne-0.16.1-rpi-test.zip](https://github.com/atauenis/webone/files/13453072/WebOne-0.16.1-rpi-test.zip).** There's a .NET 7 and .NET 8 based builds for comparison.

Tested it on my Pi 1B with Raspbian 11 Lite, it runs and working.
